### PR TITLE
Stream Layer v32

### PR DESCRIPTION
**Pull request by hpihkala**
This PR bumps `Stream Layer` version to `32`.

- Rename `Message Layer` to `Stream Layer` for clarity
- Add `groupKeyId` to all `Stream Layer` messages
- Split `contentType` to two independent fields: `messageType` (how the message should be handled) and `contentType` (how the content is encoded)
- Refactor the `GroupKey*` message types to enable subscribers to request particular `encryptionKeyId`
- Remove `encryptionType 3 (NEW_KEY_AND_AES)`, which is a hack, and add a new `messageType 32 (GroupKeyRotate)`. Where publishers would previously publish a `NEW_KEY_AND_AES` message, they should now publish two messages: a `GroupKeyRotate` (encrypted with the old group key), and a normal `StreamMessage` (encrypted with the new key)

### CODE REVIEW

File: PROTOCOL.md [hunk 1]
```diff
@@ -427,63 +427,204 @@ Field    | Type | Description
...
-`signatureType` | Name | Description | Signature payload fields to be concatenated in order
--------------- | ---- |------------ | -----------------------
-0 | `NONE` | No signature. signature field is empty in this case. | None.
-1 | `ETH_LEGACY` | Ethereum signature produced by old clients (Message Layer version 29). The signature field is encoded as a hex string. | `streamId`, `streamPartition`, `timestamp`, `publisherId`, `content`
-2 | `ETH` | Ethereum signature produced by current clients (since Message Layer version 30). The signature field is encoded as a hex string. | all the `msgId` fields, (`streamId`, `streamPartition`, `timestamp`, `sequenceNumber`, `publisherId`, `msgChainId`), `prevMsgRef`, `content`
+Describes how the content (post-decryption in case it's encrypted) should be parsed.
+
+`contentType` | Description
+------------- | -----------
+0 | JSON
```
Comment (timoxley) [inline PROTOCOL.md:482]: Great addition. 

Maybe should define these straight up using the same format as HTTP? i.e. use mime type + also define the charset used so as to avoid any ambiguity https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type

File: PROTOCOL.md [hunk 2]
```diff
@@ -427,63 +427,204 @@ Field    | Type | Description
...
 #### `encryptionType`
 
+For each non-zero `encryptionType`, the `contentType` is a hex-encoded string.
+
 `encryptionType` | Description
 -------------- | --------
 0 | Unencrypted, plaintext message.
-RSA | Content is asymetrically encrypted using RSA.
-AES | Content is symmetrically encrypted using AES.
-NEW_KEY_AND_AES | The content contains a concatenation of a new AES key and the content, encrypted with the old key.
```
Comment (timoxley) [inline PROTOCOL.md:482]: 👍

File: PROTOCOL.md (hunk 1 above)
Comment (timoxley) [reply PROTOCOL.md:482]: Also wondering if we should have an escape hatch in addition to a hard-coded list?

File: PROTOCOL.md [hunk 3]
```diff
@@ -427,63 +427,204 @@ Field    | Type | Description
...
+
+`contentType` | Description
+------------- | -----------
+0 | JSON
+
+Other content types, including binary types, will be defined in the future.
 
 #### `encryptionType`
 
+For each non-zero `encryptionType`, the `contentType` is a hex-encoded string.
```
Comment (timoxley) [inline PROTOCOL.md]: I think this is saying the implicit `contentType` pre-decryption is a hex-encoded string, which means that the content itself will be a hex-encoded string, but I had to read this a few times to get that.

File: PROTOCOL.md [hunk 4]
```diff
@@ -427,63 +427,204 @@ Field    | Type | Description
...
+
+```
+[32, [...msgIdFields], [...msgRefFields], 31, 0, 0, ["requestId", "streamId", "ERROR_CODE", "Error message"], 2, "0x29c057786Fa..."]
+```
+
+### GroupKeyRotate
+
+This message is sent by publishers to rotate the encryption key and deliver it to existing subscribers (for forward secrecy). Subscribers should add the contained key into their key storage. The message is encrypted with another (previous) key. Publishers usually send this message to inform subscribers of a new key in advance, so that they every subscriber doesn't need to send a `GroupKeyRequest` to the publisher when they encounter the first `StreamMessage` encrypted with the new key.
+
+`GroupKeyRotate`s are always encrypted (`encryptionType 2 (AES)`) with a previous key. The `content` encoded as `contentType 0` (JSON) is as follows:
```
Comment (timoxley) [inline PROTOCOL.md]: with "a previous key" i.e. any previous key, or "the previous key" i.e. the key used to encrypt the immediately proceeding message? Does it matter? Maybe should be explicit if you mean "any"

File: PROTOCOL.md (hunk 3 above)
Comment (timoxley) [inline PROTOCOL.md]: The `contentType` perhaps shouldn't be mentioned at all e.g. something like:

```suggestion
For each non-zero `encryptionType`, the `content` will be a hex-encoded binary string, encrypted using the specified `groupKeyId`.
```

File: PROTOCOL.md [hunk 5]
```diff
@@ -427,63 +427,204 @@ Field    | Type | Description
...
+`GroupKeyErrorResponse`s are always unencrypted (`encryptionType 0`). The `content` encoded as `contentType 0` (JSON) is as follows:
+
+```
+[requestId, streamId, errorCode, errorMessage]
+```
+
+ Field        | Type      | Description
+ ------------ | --------- | -----------
+ `requestId`  | `string`  | The `requestId` of the `GroupKeyRequest`, identifying which request is being responded to.
+ `streamId`   | `string`  | The stream for which a key is being delivered.
```
Comment (timoxley) [inline PROTOCOL.md:624]: perhaps should include the `groupKeyId` in the error, clients can figure this out by looking at which request corresponds to which groupKeyId, but might help debugging.

File: PROTOCOL.md [hunk 6]
```diff
@@ -504,7 +645,7 @@ Field    | Type | Description
 
 ### MessageRef
 
-Used inside a `StreamMessage` to identify the previous message on the same `msgChainId` (defined above).
+Used inside a Stream Layer message to identify the previous message on the same `msgChainId` (defined above). If the `MessageRef` is not provided by the publisher, then the stream will have weaker deliver guarantees, as subscribers will not be able to detect missing messages and request gapfills.
```
Comment (timoxley) [inline PROTOCOL.md]:
```suggestion
Used inside a Stream Layer message to identify the previous message on the same `msgChainId` (defined above). If the `MessageRef` is not provided by the publisher, then the stream will have weaker delivery guarantees, as subscribers will not be able to detect missing messages and request gapfills.
```

File: PROTOCOL.md [hunk 7]
```diff
@@ -427,63 +427,204 @@ Field    | Type | Description
...
+ 
+Example of a `GroupKeyReset` including the rest of the `Stream Layer` fields:
+
+```
+[32, [...msgIdFields], [...msgRefFields], 30, 0, 0, ["keyId", "123abc"], 2, "0x29c057786Fa..."]
+```
+
+### GroupKeyErrorResponse
+
+In case there's an error processing a `GroupKeyRequest`, this message gets sent by a publisher to the requestor's key exchange stream.
```
Comment (timoxley) [inline PROTOCOL.md]: Is this what would be sent if the `groupKeyId` doesn't exist or the user doesn't have authorisation for that key e.g. user product subscription time expired?

File: PROTOCOL.md [hunk 8]
```diff
@@ -427,63 +427,204 @@ Field    | Type | Description
...
+1 | `ETH_LEGACY` | Ethereum signature produced by old clients. The signature field is encoded as a hex string. | `streamId`, `streamPartition`, `timestamp`, `publisherId`, `content`
+2 | `ETH` | Ethereum signature produced by current clients (since Stream Layer version 30). The signature field is encoded as a hex string. | `streamId`, `streamPartition`, `timestamp`, `sequenceNumber`, `publisherId`, `msgChainId`, `prevMsgRef.timestamp`, `prevMsgRef.sequenceNumber`, `content`
+
+### StreamMessage
+
+Contains an arbitrary, application-specific `content` payload, i.e. the `content` of this message should be passed to the subscribing application's message handler. This is the message type that applications interact with via publishing and subscribing to messages.
+
+Example (unencrypted JSON message):
+```
+[32, [...msgIdFields], [...msgRefFields], 27, 0, 0, "{\"content\":42}", 2, "0x29c057786Fa..."]
```
Comment (timoxley) [inline PROTOCOL.md]: The JSON message doesn't need to be inside the `"content":` key, right? i.e. example of "content" shouldn't include a JSON object just containing a "content" key, lest people think this is part of the protocol. It isn't, is it?

**Review by timoxley — APPROVED** [review]
Looks good

File: PROTOCOL.md (hunk 1 above)
Comment (hpihkala) [reply PROTOCOL.md:482]: > use mime type + also define the charset used

Need to keep the messages as compact as possible, and having predefined numeric values seems more efficient than text strings like `application/json`. 

Charset only comes into play when converting to/from binary, which actually isn't happening here for `StreamMessage` content separately (binary conversion happens for the whole message).

This file currently defines only the messages but not their binary serialization. I added that.

Later if we switch to a binary protocol, then the `content` would of course be serialized on its own, meaning the charset needs to become explicit for the `contentType`. Perhaps then `JSON` will be redefined to `JSON_UTF8` or similar.

Comment (hpihkala) [reply PROTOCOL.md:482]: > Also wondering if we should have an escape hatch in addition to a hard-coded list?

I didn't understand this one, can you elaborate what you mean?

File: PROTOCOL.md (hunk 3 above)
Comment (hpihkala) [reply PROTOCOL.md]: Oops, should say `content` instead of `contentType`. I'll also clarify this sentence.

File: PROTOCOL.md (hunk 4 above)
Comment (hpihkala) [reply PROTOCOL.md]: I mean the immediately preceding key. Clarified

File: PROTOCOL.md (hunk 5 above)
Comment (hpihkala) [reply PROTOCOL.md:624]: The request can have multiple `groupKeyIds`. It would make sense to include in the error response the ones that errored. I'll make that change.

This actually leads us to a situation where _some_ of the requested ids may succeed and some may fail. Perhaps in this case the publisher should send both a `GroupKeyResponse` and `GroupKeyErrorResponse`. I guess that's fine?!

Comment (hpihkala) [reply PROTOCOL.md:624]: Added some text about that, and also considered the case where the publisher has many instances connected to the network and subscribed to the key exchange stream. In this case the requestor will receive many responses for the same request.

File: PROTOCOL.md (hunk 7 above)
Comment (hpihkala) [reply PROTOCOL.md]: Yes, exactly. Clarified.

File: PROTOCOL.md (hunk 8 above)
Comment (hpihkala) [reply PROTOCOL.md]: Clarified

File: PROTOCOL.md [hunk 9]
```diff
@@ -427,63 +427,204 @@ Field    | Type | Description
...
-`version` | `number` | Is currently 30.
-`msgId` | MessageID |Array representation of the `MessageID` to uniquely identify this message. 
-`prevMsgRef` | MessageRef | Array representation of the `MessageRef` of the previous message on a message chain (defined in the `msgId`). Used to detect missing messages.
-`contentType` | `number` | Determines how the content should be parsed according to the table below.
-`encryptionType` | `number` | Encryption type as defined by the table below.
-`content` | `string` | Content data of the message.
-`signatureType` | `number` | Signature type as defined by the table below.
-`signature` | `string` | Signature of the message, signed by the producer. Encoding depends on the signature type.
+ ```
+ [version, msgId, prevMsgRef, messageType, contentType, encryptionType, encryptionKeyId, content, signatureType, signature]
```
Comment (timoxley) [inline PROTOCOL.md]:
```suggestion
 [version, msgId, prevMsgRef, messageType, contentType, encryptionType, groupKeyId, content, signatureType, signature]
```

IMO we **should** rename all the `groupKey` stuff to `encryptionKey` since  "groupKey" being related to encryption at all is not immediately obvious, unlike "encryptionKey", at least to me.

Comment (timoxley) [reply PROTOCOL.md]: (also needs same change in PR description)

Comment (hpihkala) [reply PROTOCOL.md]: Yeah this crossed my mind as well. I started making the change, but then aborted. I didn't want to snowball to renaming the `GroupKeyRequest` message types as well.

I also thought that the protocol probably should not be implemented by someone who doesn't know the basics of cryptography. So keeping the `GroupKeyRequest` seems fine from that point of view.

But actually now, with the latest commit I added that the `groupKeyId` field could in case of RSA encryption be set to the recipient's public key. That conflicts with the terminology, since a group key means a shared symmetric key, so in case the field can be used to point to a RSA key pair as well, it would not be a good name for the field.

I'll have another pass and see what might work best!

Comment (hpihkala) [reply PROTOCOL.md]: I don't want to snowball this at this time, so leaving the naming as is for now. "Group key" isn't by any means incorrect, it's even "more" correct (accurate) than "encryption key". Can be reconsidered in the future.

The important thing IMO here is that we make sure that the interface of the libraries are friendly enough to be used by users not familiar with cryptography. The spec doesn't need to be written in layman terms.

The only thing that's "wrong" now is that we're "abusing" the `groupKeyId` field to identify the RSA key used in case of RSA encryption. That is not a group key. But even so, I think I'll leave it as is for now.

File: PROTOCOL.md (hunk 1 above)
Comment (timoxley) [reply PROTOCOL.md:482]: > Charset only comes into play when converting to/from binary

Charset is always in play? JSON is **supposed** to always be in UTF{8,16,32} but if an intermediary accidentally re-encodes using ISO-8859-1 or some other charset, then there may be issues decoding the text correctly.

On this note, we should probably have a test in each client & the testing-tool that confirms valid strings like "κόσμε" are decoded correctly by the clients (see https://www.w3.org/2001/06/utf-8-wrong/UTF-8-test.html and https://www.w3.org/2001/06/utf-8-test/UTF-8-demo.html), and don't get corrupted going in and out of storage.

Comment (timoxley) [reply PROTOCOL.md:482]: > having predefined numeric values seems more efficient than text strings like application/json.

Agreed, not suggesting including the entire mimetype in the protocol (but perhaps as an optional escape hatch), just suggesting we define the numeric types using mimetype + char encoding instead of just "JSON".

Comment (timoxley) [reply PROTOCOL.md:482]: Added an character encoding integration test to the JS client here, and it seems to work. 👍  https://github.com/streamr-dev/streamr-client-javascript/pull/151

Comment (hpihkala) [reply PROTOCOL.md:482]: What I meant was:

- `contentType` describes how the (text in!) `content` should be interpreted

- `content` is encoded together with the whole surrounding message (with a charset defined for the whole protocol, which is utf8)

What follows from the above is that it's not useful (or possible) for `content` to have a separate encoding.

So this is a limitation that stems from the nature of being a text-based protocol.

Sure, one could send non-utf8 content by sending arbitrary bytes in the content by encoding binary to text, but that would be wildly inefficient. And that's what we're already doing for encrypted content and signatures (and yeah it's inefficient).

So to enable full flexibility and get rid of the inefficiencies of a text protocol, in the medium term we definitely need to move to a binary-first protocol, probably protobuf-based. But first things first!

Comment (timoxley) [reply PROTOCOL.md:482]: > Sure, one could send non-utf8 content by sending arbitrary bytes in the content by encoding binary to text, but that would be wildly inefficient

I'm talking about about special/international charset support for text, not about encoding different data types. i.e. text can use utf8, uc2, utf16, ISO-8859-1 (latin1), Windows-1252, ascii, ansi, etc. Some tools (e.g. mysql) will default to latin1 (the default charset for html) or ascii if not explicitly told to use utf8.

I'm just recommending that the protocol be totally explicit about which charset should be used so nobody can complain we didn't specify a charset if/when their non-utf8 text gets mangled by an implicit conversion to utf8.

Comment (timoxley) [reply PROTOCOL.md:482]: Ahh I see now: https://github.com/streamr-dev/streamr-specs/pull/7/commits/1d8fe8b59b4bbb674b860ee9b9e5cf06e54c64b6

> The JSON messages are UTF8 encoded when transported under an underlying binary protocol (which from the perspective of the message definitions is undefined, but in practice the network uses websocket and in the future WebRTC).  

Perfect 👍
